File: 프로그래머스/2/87946. 피로도/피로도.cpp
#include <string>
#include <vector>
#include <algorithm>
#define X first
#define Y second

using namespace std;

vector<int> v;

int solution(int k, vector<vector<int>> dungeons) {
    int answer = -1;    
    
    for(int i = 0; i < (int)dungeons.size(); i++){
        v.push_back(i);
    }
    
    do{
        int cnt = 0;
        int tmp = k;
        
        for(int i = 0; i < (int)dungeons.size(); i++){
            if(tmp >= dungeons[v[i]][0]){
                tmp -= dungeons[v[i]][1]; cnt++;
            }
        }    
        
        answer = max(answer, cnt);
    } while(next_permutation(v.begin(), v.end()));
    
    return answer;
}

File: 프로그래머스/2/87946. 피로도/피로도.java
import java.util.*;

class Solution {
    static int n;
    static boolean[] vis;
    static int[] arr;
    static int mx;
    
    public int solution(int k, int[][] dungeons) {
        n = dungeons.length;
        vis = new boolean[n];
        arr = new int[n];

        backtracking(0, k, dungeons);
        
        return mx;
    }
    
    static void backtracking(int k, int cur, int[][] dungeons) {
        if(mx == n)
            return;
        if(n == k){
            int cnt = 0;
            int now = cur;
            for(int i = 0; i < n; i++){
                
                if(dungeons[arr[i]][0] <= now) {
                    now -= dungeons[arr[i]][1]; cnt++;
                }
            }
            
            System.out.println();
            
            mx = Math.max(cnt, mx);
            
            
            return;
        }
        
        for(int i = 0; i < n; i++){
            if(vis[i]) continue;
            
            arr[k] = i;
            vis[i] = true;
            
            backtracking(k + 1, cur, dungeons);
            
            vis[i] = false;
        }
    }
}